AzureMonitorAgent/ama_tst/modules/helpers.py

```diff
@@ -93,16 +111,77 @@ def find_package_manager():
                 pkg_manager = "rpm"
             except subprocess.CalledProcessError:
                 pass
-
+    general_info['PKG_MANAGER'] = pkg_manager
     return pkg_manager
 
+def get_package_version(pkg):
+    pkg_mngr = geninfo_lookup('PKG_MANAGER')
+    # dpkg
+    if (pkg_mngr == 'dpkg'):
+        return get_dpkg_pkg_version(pkg)
+    # rpm
+    elif (pkg_mngr == 'rpm'):
+        return get_rpm_pkg_version(pkg)
+    else:
+        return None
+    
+# Package Info
+def get_dpkg_pkg_version(pkg):
+    try:
+        dpkg_info = subprocess.check_output(['dpkg', '-s', pkg], universal_newlines=True,\
+                                            stderr=subprocess.STDOUT)
+        dpkg_lines = dpkg_info.split('\n')
+        for line in dpkg_lines:
+            if (line.startswith('Package: ') and not line.endswith(pkg)):
+                # wrong package
+                return None
+            if (line.startswith('Status: ') and not line.endswith('installed')):
+                # not properly installed
+                return None
+            if (line.startswith('Version: ')):
+                version = (line.split())[-1]
+                general_info['{0}_VERSION'.format(pkg.upper())] = version
+                return version
+        return None
+    except subprocess.CalledProcessError:
+        return None
+
+def get_rpm_pkg_version(pkg):
+    try:
+        rpm_info = subprocess.check_output(['rpm', '-qi', pkg], universal_newlines=True,\
+                                            stderr=subprocess.STDOUT)
+        if ("package {0} is not installed".format(pkg) in rpm_info):
+            # didn't find package
+            return None
+        rpm_lines = rpm_info.split('\n')
+        for line in rpm_lines:
+            # parse line
+            # note: parsing is weird bc rpm puts various information into two columns
```

Comment:
JFYI, the logic looks fine in any case, but the output looks different on different distros evidently, e.g. my centos7 machine:

```
[root@hestolz-centos-arc situhan]# rpm -qi azuremonitoragent
Name        : azuremonitoragent
Version     : 1.22.1
Release     : build.master.443
Architecture: x86_64
Install Date: Fri Sep  9 08:28:49 2022
Group       : Productivity/Monitoring
Size        : 112077757
License     : MICROSOFT SOFTWARE LICENSE TERMS
Signature   : RSA/SHA256, Fri Sep  2 20:29:43 2022, Key ID eb3e94adbe1229cf
Source RPM  : azuremonitoragent-1.22.1-build.master.443.src.rpm
Build Date  : Fri Sep  2 19:27:24 2022
Build Host  : 45ac5e19c8d9
Relocations : (not relocatable)
URL         : https://aka.ms/azuremonitor
Summary     : Azure Monitor Agent daemon
Description :
Azure Monitor Agent daemon
```

Lets sanity check that there aren't any issues parsing rpm or dpkg output depending on the version of the package manager itself

Reply:
I checked a few distros and it looks like the latest rpm/dpkg output can be parsed in this way. If the output cannot be parsed it will be displayed in errors. For sanity check, is there any other checks I can do before parsing?

Reply:
I think its fine. If we checked on a few distros and its generally consistent and parseable and any errors would get surfaced should be sufficient.